File: 2015/lua/5/main.lua
function check_good_string_part1(str)
    local vowels = 0
    local double = false
    local bad = false

    local last = ""
    for i = 1, #str do
        local c = str:sub(i, i)
        if c == "a" or c == "e" or c == "i" or c == "o" or c == "u" then
            vowels = vowels + 1
        end

        if c == last then
            double = true
        end

        if last == "a" and c == "b" then
            bad = true
        end
        if last == "c" and c == "d" then
            bad = true
        end
        if last == "p" and c == "q" then
            bad = true
        end
        if last == "x" and c == "y" then
            bad = true
        end

        last = c
    end

    return vowels >= 3 and double and not bad
end

function check_good_string_part2(str)
    local double = false
    local repeat_ = false

    for i = 1, #str - 1 do
        local pair = str:sub(i, i + 1)
        if str:find(pair, i + 2) then
            repeat_ = true
        end

        if str:sub(i, i) == str:sub(i + 2, i + 2) then
            double = true
        end
    end

    return double and repeat_
end

function main()
    -- get command line arguments
    if #arg < 1 then
        print("Usage: lua main.lua <input.txt>")
        return
    end

    -- read input file
    local input = arg[1]
    local file = io.open(input, "r")
    if not file then
        print("Error: could not open file " .. input)
        return
    end

    -- read file line by line
    local lines = {}
    for line in file:lines() do
        table.insert(lines, line)
    end

    local part1_count = 0
    local part2_count = 0

    for i, line in ipairs(lines) do
        if check_good_string_part1(line) then
            part1_count = part1_count + 1
        end

        if check_good_string_part2(line) then
            part2_count = part2_count + 1
        end
    end

    print("Part 1: " .. part1_count)
    print("Part 2: " .. part2_count)
end

main()

File: 2015/rust/5/src/main.rs
fn main() {
    // get command line arguments
    let args: Vec<String> = std::env::args().collect();
    // check if there are enough arguments
    if args.len() < 2 {
        // print usage
        println!("Usage: cargo run <input.txt>");
        // exit with error
        std::process::exit(1);
    }

    // read the file
    let contents = std::fs::read_to_string(&args[1]).expect("Failed to read file");
    // print the file contents
    let mut part1_count = 0;
    let mut part2_count = 0;
    for line in contents.lines() {
        if check_good_string_part1(&line.to_string()) {
            part1_count += 1;
        }

        if check_good_string_part2(&line.to_string()) {
            part2_count += 1;
        }
    }

    println!("Part 1: {}", part1_count);
    println!("Part 2: {}", part2_count);
}

fn check_good_string_part1(s: &String) -> bool {
    let mut vowels = 0;
    let mut double = false;
    let mut bad = false;
    let mut last = ' ';
    for c in s.chars() {
        if c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' {
            vowels += 1;
        }
        if c == last {
            double = true;
        }
        if (last == 'a' && c == 'b')
            || (last == 'c' && c == 'd')
            || (last == 'p' && c == 'q')
            || (last == 'x' && c == 'y')
        {
            bad = true;
        }
        last = c;
    }
    vowels >= 3 && double && !bad
}

fn check_good_string_part2(s: &String) -> bool {
    let mut double = false;
    let mut repeat = false;
    let mut last = ' ';
    let mut last_last = ' ';
    for c in s.chars() {
        if c == last_last {
            repeat = true;
        }
        if last != ' ' {
            let pair = format!("{}{}", last, c);
            if s.matches(&pair).count() > 1 {
                double = true;
            }
        }
        last_last = last;
        last = c;
    }
    double && repeat
}
